### PR TITLE
Personal/macarl/gatedbuild

### PR DESCRIPTION
**Pull request by matthewfcarlson**
Added configuration validation as well as added skipping tests as needed

### CODE REVIEW

File: UefiBuild/MuBuild/ConfigValidator.py [hunk 1]
```diff
@@ -0,0 +1,307 @@
+## @file ConfigValidator.py
+# This module contains support for Outputting Junit xml.
```
Comment (spbrogan) [inline UefiBuild/MuBuild/ConfigValidator.py]: comment incorrect

File: UefiBuild/MuBuild/MuBuild.py [hunk 2]
```diff
@@ -70,9 +72,63 @@ def get_mu_config():
     parser.add_argument (
     '-p', '--pkg','--pkg-dir', dest = 'pkg', required = False, type=str,help = 'The package or folder you want to test/compile relative to the Mu Config'
     )
+    parser.add_argument (
+    '-check-config', '--cc', dest = 'check_config', required = False, action="store_true",help = 'The package or folder you want to test/compile relative to the Mu Config'
```
Comment (spbrogan) [inline UefiBuild/MuBuild/MuBuild.py]: incorrect

File: UefiBuild/PluginManager.py [hunk 3]
```diff
@@ -215,6 +215,9 @@ def GetTestName(self, packagename, environment):
     def IsTargetDependent(self):
         return False
 
+    def ValidateConfig(self,config):
```
Comment (spbrogan) [inline UefiBuild/PluginManager.py]: what does a plugin do if error.

File: UefiBuild/MuBuild/MuBuild.py [hunk 4]
```diff
@@ -191,17 +252,21 @@ def get_mu_config():
...
-        for Descriptor in pluginManager.GetPluginsOfClass(PluginManager.IMuBuildPlugin):
+        for Descriptor in pluginList:
             #Get our targets
             targets = ["DEBUG"]
             if Descriptor.Obj.IsTargetDependent() and "Targets" in mu_config:
                 targets = mu_config["Targets"]
 
-
+            #merge the repo level and package level
+            pkg_configuration = merge_config(mu_config,pkg_config,Descriptor.descriptor)
```
Comment (spbrogan) [inline UefiBuild/MuBuild/MuBuild.py]: shouldn't this happen before check_package_config?  
Does it really need to happen on this inner loop or could it happen 1 loop up?

**Review by spbrogan — CHANGES_REQUESTED** [review]

File: UefiBuild/PluginManager.py (hunk 3 above)
Comment (matthewfcarlson) [reply UefiBuild/PluginManager.py]: Generally it's expected that a plugin will throw an exception if it encounters an error just like the validators

**Comment by matthewfcarlson** [comment]
Fixed all the issues except the validate config

File: UefiBuild/PluginManager.py (hunk 3 above)
Comment (matthewfcarlson) [reply UefiBuild/PluginManager.py]: Another solution would be to return false or some message about what went wrong but then it's harder to see where the exception is coming from

**Comment by matthewfcarlson** [comment]
I've removed the case insensitive dictionary and now I've fixed a few other things I noticed

**Review by spbrogan — CHANGES_REQUESTED** [review]
looks like you pulled in a submodule.  can you remove that

**Comment by matthewfcarlson** [comment]
Removed the submodule

**Comment by spbrogan** [comment]
another one in common

**Review by spbrogan — APPROVED** [review]

**Pull request merged**